File: qmd-go/pkg/store/chunker.go
package store

import (
	"unicode/utf8"
)

// Chunk represents a document chunk
type Chunk struct {
	Seq int
	Pos int
	Text string
}

const (
	DefaultChunkSize = 3200
	DefaultOverlap   = 480
)

// ChunkDocument splits a document into chunks
func ChunkDocument(text string, chunkSize, overlap int) []Chunk {
	if chunkSize <= 0 {
		chunkSize = DefaultChunkSize
	}
	if overlap <= 0 {
		overlap = DefaultOverlap
	}

	if len(text) <= chunkSize {
		return []Chunk{{Seq: 0, Pos: 0, Text: text}}
	}

	var chunks []Chunk
	pos := 0
	seq := 0

	for pos < len(text) {
		end := pos + chunkSize
		if end > len(text) {
			end = len(text)
		}

		// Try to break at sentence boundary
		if end < len(text) {
			// Look for sentence endings: . ! ? followed by space or newline
			for i := end - 1; i > end-200 && i > pos; i-- {
				if text[i] == '.' || text[i] == '!' || text[i] == '?' {
					if i+1 < len(text) && (text[i+1] == ' ' || text[i+1] == '\n') {
						end = i + 1
						break
					}
				}
			}
		}

		// Ensure we don't break in the middle of a UTF-8 character
		for !utf8.RuneStart(text[end]) && end > pos {
			end--
		}

		chunk := Chunk{
			Seq: seq,
			Pos: pos,
			Text: text[pos:end],
		}
		chunks = append(chunks, chunk)

		seq++
		pos = end - overlap
		if pos <= 0 {
			break
		}
	}

	return chunks
}

// CountTokens estimates token count (approximate: 1 token ≈ 4 chars)
func CountTokens(text string) int {
	return len(text) / 4
}

File: qmd-python/src/qmd/store/chunker.py
"""Document chunking utilities"""

from dataclasses import dataclass


@dataclass
class Chunk:
    """Represents a document chunk"""
    seq: int
    pos: int
    text: str


DEFAULT_CHUNK_SIZE = 3200  # ~800 tokens
DEFAULT_OVERLAP = 480      # ~120 tokens


def chunk_document(text: str, chunk_size: int = DEFAULT_CHUNK_SIZE, overlap: int = DEFAULT_OVERLAP) -> list[Chunk]:
    """Split a document into overlapping chunks"""
    if len(text) <= chunk_size:
        return [Chunk(seq=0, pos=0, text=text)]

    chunks = []
    pos = 0
    seq = 0

    while pos < len(text):
        end = pos + chunk_size
        if end > len(text):
            end = len(text)

        # Try to break at sentence boundary
        if end < len(text):
            # Look for sentence endings: . ! ? followed by space or newline
            for i in range(end - 1, max(end - 200, pos), -1):
                if text[i] in '.!?' and i + 1 < len(text) and text[i + 1] in ' \n':
                    end = i + 1
                    break

        chunk = Chunk(seq=seq, pos=pos, text=text[pos:end])
        chunks.append(chunk)

        seq += 1
        pos = end - overlap
        if pos <= 0:
            break

    return chunks


def count_tokens(text: str) -> int:
    """Estimate token count (approximate: 1 token ≈ 4 chars)"""
    return len(text) // 4
